### PR TITLE
[MCC-412658] Add V2 to benchmark

### PR DESCRIPTION
**Pull request by cmcinnes-mdsol**
@mdsol/team-16 this adds V2 signing to the benchmark rake task.

Adding DNM because this is pointed at the `feature/v2-mauth-protocol`.

### CODE REVIEW

**Comment by cmcinnes-mdsol** [comment]
```
Warming up --------------------------------------
       v1-sign-short    89.000  i/100ms
       v2-sign-short    94.000  i/100ms
     both-sign-short    51.000  i/100ms
     v1-sign-average    87.000  i/100ms
     v2-sign-average    87.000  i/100ms
   both-sign-average    43.000  i/100ms
        v1-sign-huge     6.000  i/100ms
        v2-sign-huge     7.000  i/100ms
      both-sign-huge     3.000  i/100ms
Calculating -------------------------------------
       v1-sign-short      1.067k (± 4.6%) i/s -      5.340k in   5.014668s
       v2-sign-short      1.050k (± 2.7%) i/s -      5.264k in   5.014746s
     both-sign-short    526.527  (± 4.4%) i/s -      2.652k in   5.047097s
     v1-sign-average    919.132  (± 6.7%) i/s -      4.611k in   5.040822s
     v2-sign-average    954.075  (± 3.0%) i/s -      4.785k in   5.020007s
   both-sign-average    481.066  (± 3.5%) i/s -      2.408k in   5.011711s
        v1-sign-huge     66.967  (± 3.0%) i/s -    336.000  in   5.020862s
        v2-sign-huge     72.196  (± 4.2%) i/s -    364.000  in   5.051811s
      both-sign-huge     35.150  (± 2.8%) i/s -    177.000  in   5.044630s

Comparison:
       v1-sign-short:     1067.3 i/s
       v2-sign-short:     1050.5 i/s - same-ish: difference falls within error
     v2-sign-average:      954.1 i/s - 1.12x  slower
     v1-sign-average:      919.1 i/s - 1.16x  slower
     both-sign-short:      526.5 i/s - 2.03x  slower
   both-sign-average:      481.1 i/s - 2.22x  slower
        v2-sign-huge:       72.2 i/s - 14.78x  slower
        v1-sign-huge:       67.0 i/s - 15.94x  slower
      both-sign-huge:       35.2 i/s - 30.36x  slower

i/s means the number of signatures of a message per second.


Warming up --------------------------------------
v1-authenticate-short
                         1.001k i/100ms
v2-authenticate-short
                         1.008k i/100ms
v1-authenticate-average
                       523.000  i/100ms
v2-authenticate-average
                       522.000  i/100ms
v1-authenticate-huge     7.000  i/100ms
v2-authenticate-huge     8.000  i/100ms
Calculating -------------------------------------
v1-authenticate-short
                         10.187k (± 3.4%) i/s -     51.051k in   5.017151s
v2-authenticate-short
                         10.661k (± 6.0%) i/s -     53.424k in   5.030617s
v1-authenticate-average
                          5.630k (± 2.9%) i/s -     28.242k in   5.020521s
v2-authenticate-average
                          5.675k (± 3.1%) i/s -     28.710k in   5.064085s
v1-authenticate-huge     84.891  (± 3.5%) i/s -    427.000  in   5.036918s
v2-authenticate-huge     81.303  (± 2.5%) i/s -    408.000  in   5.022799s

Comparison:
v2-authenticate-short:    10660.9 i/s
v1-authenticate-short:    10187.4 i/s - same-ish: difference falls within error
v2-authenticate-average:     5675.1 i/s - 1.88x  slower
v1-authenticate-average:     5630.3 i/s - 1.89x  slower
v1-authenticate-huge:       84.9 i/s - 125.58x  slower
v2-authenticate-huge:       81.3 i/s - 131.13x  slower

i/s means the number of authentication checks of signatures per second.
```

File: Rakefile [hunk 1]
```diff
@@ -48,26 +60,42 @@ task :benchmark do
   average_request = TestSignableRequest.new(verb: 'PUT', request_url: '/', body: average_body)
   huge_request = TestSignableRequest.new(verb: 'PUT', request_url: '/', body: huge_body)
```
Comment (ejinotti-mdsol) [inline Rakefile]: Perhaps we should have some reqs with query parameters for V2 as well?

Comment (cmcinnes-mdsol) [reply Rakefile]: Added. Thought about adding a normal size query string and a huge query string but a huge query string doesn't seem very likely (whereas a huge body seems relatively likely).

**Comment by cmcinnes-mdsol** [comment]
Results with request with query params 
```
Warming up --------------------------------------
       v1-sign-short   111.000  i/100ms
       v2-sign-short   110.000  i/100ms
     both-sign-short    56.000  i/100ms
          v1-sign-qs   113.000  i/100ms
          v2-sign-qs   111.000  i/100ms
        both-sign-qs    55.000  i/100ms
     v1-sign-average    96.000  i/100ms
     v2-sign-average   100.000  i/100ms
   both-sign-average    49.000  i/100ms
        v1-sign-huge     6.000  i/100ms
        v2-sign-huge     6.000  i/100ms
      both-sign-huge     3.000  i/100ms
Calculating -------------------------------------
       v1-sign-short      1.088k (± 2.1%) i/s -      5.550k in   5.101938s
       v2-sign-short      1.096k (± 1.8%) i/s -      5.500k in   5.018159s
     both-sign-short    550.953  (± 2.4%) i/s -      2.800k in   5.084961s
          v1-sign-qs      1.099k (± 2.1%) i/s -      5.537k in   5.038208s
          v2-sign-qs      1.071k (± 2.1%) i/s -      5.439k in   5.082612s
        both-sign-qs    535.304  (± 3.4%) i/s -      2.695k in   5.041466s
     v1-sign-average    939.048  (± 3.5%) i/s -      4.704k in   5.015858s
     v2-sign-average    942.249  (± 4.9%) i/s -      4.700k in   5.000985s
   both-sign-average    493.202  (± 2.0%) i/s -      2.499k in   5.068926s
        v1-sign-huge     66.597  (± 1.5%) i/s -    336.000  in   5.046826s
        v2-sign-huge     75.718  (± 1.3%) i/s -    384.000  in   5.073196s
      both-sign-huge     35.947  (± 2.8%) i/s -    180.000  in   5.010271s

Comparison:
          v1-sign-qs:     1099.5 i/s
       v2-sign-short:     1096.4 i/s - same-ish: difference falls within error
       v1-sign-short:     1088.3 i/s - same-ish: difference falls within error
          v2-sign-qs:     1070.6 i/s - same-ish: difference falls within error
     v2-sign-average:      942.2 i/s - 1.17x  slower
     v1-sign-average:      939.0 i/s - 1.17x  slower
     both-sign-short:      551.0 i/s - 2.00x  slower
        both-sign-qs:      535.3 i/s - 2.05x  slower
   both-sign-average:      493.2 i/s - 2.23x  slower
        v2-sign-huge:       75.7 i/s - 14.52x  slower
        v1-sign-huge:       66.6 i/s - 16.51x  slower
      both-sign-huge:       35.9 i/s - 30.59x  slower

i/s means the number of signatures of a message per second.


Warming up --------------------------------------
v1-authenticate-short
                         1.071k i/100ms
v2-authenticate-short
                         1.100k i/100ms
  v1-authenticate-qs     1.075k i/100ms
  v2-authenticate-qs     1.081k i/100ms
v1-authenticate-average
                       531.000  i/100ms
v2-authenticate-average
                       536.000  i/100ms
v1-authenticate-huge     7.000  i/100ms
v2-authenticate-huge     7.000  i/100ms
Calculating -------------------------------------
v1-authenticate-short
                         10.946k (± 3.2%) i/s -     55.692k in   5.093447s
v2-authenticate-short
                         10.973k (± 3.0%) i/s -     55.000k in   5.016782s
  v1-authenticate-qs     10.812k (± 3.7%) i/s -     54.825k in   5.078050s
  v2-authenticate-qs     10.375k (± 5.5%) i/s -     51.888k in   5.018736s
v1-authenticate-average
                          5.117k (± 6.4%) i/s -     26.019k in   5.107858s
v2-authenticate-average
                          5.275k (± 5.0%) i/s -     26.800k in   5.096239s
v1-authenticate-huge     66.581  (±13.5%) i/s -    329.000  in   5.041414s
v2-authenticate-huge     76.575  (± 9.1%) i/s -    385.000  in   5.079093s

Comparison:
v2-authenticate-short:    10973.2 i/s
v1-authenticate-short:    10945.6 i/s - same-ish: difference falls within error
  v1-authenticate-qs:    10811.9 i/s - same-ish: difference falls within error
  v2-authenticate-qs:    10375.4 i/s - same-ish: difference falls within error
v2-authenticate-average:     5274.8 i/s - 2.08x  slower
v1-authenticate-average:     5117.4 i/s - 2.14x  slower
v2-authenticate-huge:       76.6 i/s - 143.30x  slower
v1-authenticate-huge:       66.6 i/s - 164.81x  slower

i/s means the number of authentication checks of signatures per second.
```

File: Rakefile (hunk 1 above)
Comment (ejinotti-mdsol) [reply Rakefile]: Don't think we need the V1 with query case tho, since V1 does not do anything with query 😉

Comment (cmcinnes-mdsol) [reply Rakefile]: Removed

File: Rakefile [hunk 2]
```diff
@@ -44,30 +56,54 @@ task :benchmark do
...
+
+  v2_short_signed_request = mc.signed_v2(short_request)
+  v2_qs_signed_request = mc.signed_v1(qs_request)
+  v2_average_signed_request = mc.signed_v2(average_request)
+  v2_huge_signed_request = mc.signed_v1(huge_request)
+
   short_signed_request = mc.signed(short_request)
+  qs_signed_request = mc.signed(qs_request)
   average_signed_request = mc.signed(average_request)
   huge_signed_request = mc.signed(huge_request)
```
Comment (ejennings-mdsol) [inline Rakefile]: Can you add a `puts` that tersely explains what short, average, huge, and qs mean?

Comment (cmcinnes-mdsol) [reply Rakefile]: @ejennings-mdsol added.

File: Rakefile [hunk 3]
```diff
@@ -26,11 +26,23 @@ class TestSignableRequest < MAuth::Request
...
 task :benchmark do
-  mc = MAuth::Client.new(private_key: OpenSSL::PKey::RSA.generate(2048), app_uuid: SecureRandom.uuid)
+  mc = MAuth::Client.new(
+    private_key: OpenSSL::PKey::RSA.generate(2048),
+    app_uuid: SecureRandom.uuid,
+    v2_only_sign_requests: false
+  )
   authenticating_mc = MAuth::Client.new(mauth_baseurl: 'http://whatever', mauth_api_version: 'v1')
 
   stubs = Faraday::Adapter::Test::Stubs.new
```
Comment (ejennings-mdsol) [inline Rakefile]: Sorry, not you, but can you fix: `DEPRECATION: Using `stub` from rspec-mocks' old `:should` syntax without explicitly enabling the syntax is deprecated. Use the new `:expect` syntax or explicitly enable `:should` instead. Called from /Users/ejennings/dev/mauth-client-ruby/Rakefile:53:in `block in <top (required)>'.`

File: Rakefile [hunk 4]
```diff
@@ -44,30 +56,63 @@ task :benchmark do
...
+  v1_short_signed_request = mc.signed_v1(short_request)
+  v1_average_signed_request = mc.signed_v1(average_request)
+  v1_huge_signed_request = mc.signed_v1(huge_request)
+
+  v2_short_signed_request = mc.signed_v2(short_request)
+  v2_qs_signed_request = mc.signed_v1(qs_request)
+  v2_average_signed_request = mc.signed_v2(average_request)
+  v2_huge_signed_request = mc.signed_v1(huge_request)
+
   short_signed_request = mc.signed(short_request)
```
Comment (ejennings-mdsol) [inline Rakefile]: These 4 lines aren't used any more

File: Rakefile [hunk 5]
```diff
@@ -44,30 +56,63 @@ task :benchmark do
...
+  qs = 'don=quixote&quixote=don'
+
+  puts <<-MSG
+
+    A short request has a body of 60 chars.
+    An average request has a body of 60,000 chars.
+    A huge request has a body of 6,000,000 chars.
+    A qs request has a body of 60 chars and a query string with two k/v pairs.
+    
+    MSG
```
Comment (ejennings-mdsol) [inline Rakefile]: Nitpick: remove indentation on `MSG`

**Comment by jcarres-mdsol** [comment]
Good results, it seems v2 has the same performance than v1. No regresions!

Signing seems in general pretty slow. For a no-fair comparison, the logging library logs 60 messages in the time it takes to do one sign, and we are committing sync IO to the hard-drive.

**Pull request merged**